www/makefsdata.js

```diff
@@ -1,41 +1,265 @@
+		if (file.isDirectory()) {
+			results = results.concat(getFiles(file.path));
+		} else if (file.isFile()) {
+			results.push(file.path);
+		}
+	}
+	return results;
+}
+
+function getLowerCaseFileExtension(file) {
+	const ext = file.split('.').pop();
+	return ext ? ext.toLowerCase() : '';
+}
+
+function fixFilenameForC(filename) {
+	let varName = '';
+	for (let i = 0; i < filename.length; i++) {
+		const c = filename[i];
+		if (c >= 'a' && c <= 'z') {
+			varName += c;
+		} else if (c >= 'A' && c <= 'Z') {
+			varName += c;
+		} else if (c >= '0' && c <= '9') {
+			varName += c;
+		} else if (c === '_') {
+			varName += c;
+		} else {
+			varName += '_';
+		}
+	}
+	return varName;
+}
+
+function CStringLength(str) {
+	return Buffer.from(str, 'utf8').length;
+}
+
+function createHexString(value, prependComment = false) {
+	let column = 0;
+	let hexString = '';
+	if (typeof value === 'string') {
+		const bytes = Buffer.from(value, 'utf8');
+		if (prependComment) {
+			hexString += `/* "${value}" (${bytes.length} bytes) */\n`;
+		}
+		bytes.forEach((byte) => {
+			hexString += '0x' + byte.toString(16).padStart(2, '0') + ',';
+			if (++column >= hexBytesPerLine) {
+				hexString += '\n';
+				column = 0;
+			}
+		});
+	} else if (Buffer.isBuffer(value) || value instanceof Uint8Array) {
+		value.forEach((byte) => {
+			hexString += '0x' + byte.toString(16).padStart(2, '0') + ',';
+			if (++column >= hexBytesPerLine) {
+				hexString += '\n';
+				column = 0;
+			}
+		});
+	} else {
+		assert(false, 'Invalid value type');
+	}
+
+	return hexString + '\n';
 }
 
 function makefsdata() {
-      execFile(path.normalize(process.platform !== "darwin" ? `${root}/tools/makefsdata` : `${root}/tools/makefsdata.darwin`), [path.normalize(`${rootwww}/build`), '-defl:10', '-xc:png,json', `-f:`+ path.normalize(`${root}/lib/httpd/fsdata.c`)], function(error, data) {
-        if (error) {
-            console.error(error);
-        } else {
-            correctincludes();
-        }
-    });  
+	let fsdata = '';
+	fsdata += '#include "fsdata.h"\n';
+	fsdata += '\n';
+	fsdata += '#define file_NULL (struct fsdata_file *) NULL\n';
+	fsdata += '\n';
+	fsdata += '#ifndef FS_FILE_FLAGS_HEADER_INCLUDED\n';
+	fsdata += '#define FS_FILE_FLAGS_HEADER_INCLUDED 1\n';
+	fsdata += '#endif\n';
+	fsdata += '#ifndef FS_FILE_FLAGS_HEADER_PERSISTENT\n';
+	fsdata += '#define FS_FILE_FLAGS_HEADER_PERSISTENT 0\n';
+	fsdata += '#endif\n';
+	fsdata += '/* FSDATA_FILE_ALIGNMENT: 0=off, 1=by variable, 2=by include */\n';
+	fsdata += '#ifndef FSDATA_FILE_ALIGNMENT\n';
+	fsdata += '#define FSDATA_FILE_ALIGNMENT 0\n';
+	fsdata += '#endif\n';
+	fsdata += '#ifndef FSDATA_ALIGN_PRE\n';
+	fsdata += '#define FSDATA_ALIGN_PRE\n';
+	fsdata += '#endif\n';
+	fsdata += '#ifndef FSDATA_ALIGN_POST\n';
+	fsdata += '#define FSDATA_ALIGN_POST\n';
+	fsdata += '#endif\n';
+	fsdata += '#if FSDATA_FILE_ALIGNMENT==2\n';
+	fsdata += '#include "fsdata_alignment.h"\n';
+	fsdata += '#endif\n';
+	fsdata += '\n';
```

Comment:
Wanted to mention template literals or tagged templates as alternatives to string concatenation.
They allow multi-line strings and string interpolation.

Example of template literal
```javascript
let fsdata = `#include \"fsdata.h\"

#define file_NULL (struct fsdata_file *) NULL

#ifndef FS_FILE_FLAGS_HEADER_INCLUDED
#define FS_FILE_FLAGS_HEADER_INCLUDED 1
#endif
#ifndef FS_FILE_FLAGS_HEADER_PERSISTENT
#define FS_FILE_FLAGS_HEADER_PERSISTENT 0
#endif
/* FSDATA_FILE_ALIGNMENT: 0=off, 1=by variable, 2=by include */
#ifndef FSDATA_FILE_ALIGNMENT
#define FSDATA_FILE_ALIGNMENT 0
#endif
#ifndef FSDATA_ALIGN_PRE
#define FSDATA_ALIGN_PRE
#endif
#ifndef FSDATA_ALIGN_POST
#define FSDATA_ALIGN_POST
#endif
#if FSDATA_FILE_ALIGNMENT==2
#include \"fsdata_alignment.h\"
#endif

`
```

Reply:
I considered this, but ultimately decided against it because I wanted to have explicit control over the type of line-endings used. I don't want the line-endings of the source file to "pollute" the generated file. Depending on someone's Git configuration, they might convert line endings to CRLF on checkout, which would then generate `fsdata.c` files with mixed line endings.

Reply:
Oh that makes sense, clever!